server/util/grpc_server/grpc_server.go

```diff
@@ -2,21 +2,124 @@ package grpc_server
 	"go.opentelemetry.io/otel/attribute"
 	"go.opentelemetry.io/otel/trace"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/keepalive"
+	"google.golang.org/grpc/reflection"
 
+	apipb "github.com/buildbuddy-io/buildbuddy/proto/api/v1"
+	bbspb "github.com/buildbuddy-io/buildbuddy/proto/buildbuddy_service"
+	hlpb "github.com/buildbuddy-io/buildbuddy/proto/health"
 	grpc_prometheus "github.com/grpc-ecosystem/go-grpc-prometheus"
+	_ "google.golang.org/grpc/encoding/gzip" // imported for side effects; DO NOT REMOVE.
 )
 
+var (
+	gRPCOverHTTPPortEnabled = flag.Bool("app.grpc_over_http_port_enabled", false, "Cloud-Only")
+	// Support large BEP messages: https://github.com/bazelbuild/bazel/issues/12050
+	gRPCMaxRecvMsgSizeBytes = flag.Int("app.grpc_max_recv_msg_size_bytes", 50000000, "Configures the max GRPC receive message size [bytes]")
+
+	gRPCPort  = flag.Int("grpc_port", 1985, "The port to listen for gRPC traffic on")
+	gRPCSPort = flag.Int("grpcs_port", 1986, "The port to listen for gRPCS traffic on")
+)
+
+func RegisterGRPCServer(env environment.Env) error {
+	grpcServer, err := NewGRPCServer(env, *gRPCPort, nil)
+	if err != nil {
+		return err
+	}
+	env.SetGRPCServer(grpcServer)
+	if *gRPCOverHTTPPortEnabled {
+		env.SetMux(&gRPCMux{
+			env.GetMux(),
+			grpcServer,
+		})
+	}
+	return nil
+}
+
+func RegisterGRPCSServer(env environment.Env) error {
+	if !env.GetSSLService().IsEnabled() {
+		return nil
+	}
+	creds, err := env.GetSSLService().GetGRPCSTLSCreds()
+	if err != nil {
+		return status.InternalErrorf("Error getting SSL creds: %s", err)
+	}
+	grpcsServer, err := NewGRPCServer(env, *gRPCSPort, grpc.Creds(creds))
+	if err != nil {
+		return err
+	}
+	env.SetGRPCSServer(grpcsServer)
+	return nil
+}
+
+func NewGRPCServer(env environment.Env, port int, credentialOption grpc.ServerOption) (*grpc.Server, error) {
+	// Initialize our gRPC server (and fail early if that doesn't happen).
+	hostAndPort := fmt.Sprintf("%s:%d", env.GetListenAddr(), port)
+
+	lis, err := net.Listen("tcp", hostAndPort)
+	if err != nil {
+		return nil, status.InternalErrorf("Failed to listen: %s", err)
+	}
+
+	grpcOptions := CommonGRPCServerOptions(env)
+	if credentialOption != nil {
+		grpcOptions = append(grpcOptions, credentialOption)
+		log.Printf("gRPCS listening on http://%s", hostAndPort)
+	} else {
+		log.Printf("gRPC listening on http://%s", hostAndPort)
+	}
+
+	grpcServer := grpc.NewServer(grpcOptions...)
+
+	// Support reflection so that tools like grpc-cli (aka stubby) can
+	// enumerate our services and call them.
+	reflection.Register(grpcServer)
+
+	// Support prometheus grpc metrics.
+	grpc_prometheus.Register(grpcServer)
+	// Enable prometheus latency metrics too.
+	grpc_prometheus.EnableHandlingTimeHistogram()
+
+	// Start Build-Event-Protocol and Remote-Cache services.
+	if err := build_event_server.Register(env, grpcServer); err != nil {
+		return nil, err
+	}
+	bbspb.RegisterBuildBuddyServiceServer(grpcServer, env.GetBuildBuddyServer())
```

Comment:
What's the motivation for moving the instantiation of specific services from libmain to this file? 
This file is a utility package (it's under `server/util`) for server support code. The previous location of this code seems more appropriate.

Reply:
The motivation is that I would like the grpc Register function to handle actually starting the grpc server, but we can't call `Serve` until all these related services have been registered. I think maybe the right thing to do would be to instantiate them before registering the grpc_server, store them in the env, and then register them?

Reply:
Oh, except we need one of each for the secure or not secure grpc_server, that's not great. Hmm.

Reply:
Actually, wait, we use the same buildbuddyServer for both, so presumably it is okay to share instantiations of services between the grpc and grpcs Server

Reply:
Put all the registration buts in main now ^-^